README.md
# 🧰 Fabulously Optimized · Vanilla Installer
<!-- ALL-CONTRIBUTORS-BADGE:START - Do not remove or modify this section -->
[![All Contributors](https://img.shields.io/badge/all_contributors-7-orange.svg?style=flat-square)](#contributors-)
<!-- ALL-CONTRIBUTORS-BADGE:END -->

[![CodeQL](https://github.com/Fabulously-Optimized/vanilla-installer/actions/workflows/codeql.yml/badge.svg)](https://github.com/Fabulously-Optimized/vanilla-installer/actions/workflows/codeql.yml)
[![Lint](https://github.com/Fabulously-Optimized/vanilla-installer/actions/workflows/lint.yml/badge.svg)](https://github.com/Fabulously-Optimized/vanilla-installer/actions/workflows/lint.yml)
[![Release](https://github.com/Fabulously-Optimized/vanilla-installer/actions/workflows/release.yml/badge.svg)](https://github.com/Fabulously-Optimized/vanilla-installer/actions/workflows/release.yml)

## Download

<a target="_blank" href="https://github.com/Fabulously-Optimized/vanilla-installer/releases/latest"><img alt="Download on GitHub" height="40" src="https://cdn.jsdelivr.net/npm/@intergrav/devins-badges@2/assets/compact/available/github_vector.svg"></a>
<!-- commented out until we actually publish, if at all
<a href="https://pypi.org/project/vanilla-installer"><img alt="Download from PyPI" height="40" src="https://cdn.jsdelivr.net/npm/@intergrav/devins-badges@2/assets/compact/available/pypi_vector.svg"></a>
-->

## Need help?

<a target="_blank" href="https://discord.gg/yxaXtaQqdB"><img alt="Chat with us on Discord!" height="40" src="https://cdn.jsdelivr.net/npm/@intergrav/devins-badges@2/assets/compact/social/discord-plural_vector.svg"></a>
<a target="_blank" href="https://fabulously-optimized.gitbook.io/vanilla-installer/"><img alt="Read the documentation!" height="40" src="https://cdn.jsdelivr.net/npm/@intergrav/devins-badges@2/assets/compact/documentation/gitbook_vector.svg"></a>
<a href="https://github.com/Fabulously-Optimized/issues/new/choose"><img alt="Report an issue!" height="40" src="https://cdn.jsdelivr.net/npm/@intergrav/devins-badges@2/assets/compact/social/github-plural_vector.svg">

## Abbreviations

**FO** - Fabulously Optimized

**VI** - Vanilla Installer

## Installation

### Binaries

These are quite simple to use and should be stable.
Just [head to the **Releases** tab](https://github.com/Fabulously-Optimized/vanilla-installer/releases/latest) and download the correct file for your operating system.
However, you have to manually install updates.

**Requirements**: Ability to download a file and double-click it.

**Downsides**: No easy updating

**Recommended for**: People that aren't too well versed in technology, and probably don't care about updating VI.

<!-- commented out until released
### PyPi

These binaries are part of Python's package index (abbreviated PyPi), which can then be installed via the package manager `pip`. Because of this, you can simply run `pip install vanilla-installer` to install it and `pip install --upgrade vanilla-installer` to update it.

**Requirements**: [Python](https://python.org) installed and access to a command line (Command Prompt, PowerShell, or Windows Terminal on Windows, Terminal on macOS, whatever your desktop environment provides on Linux (e.g. GNOME Terminal, Konsole, etc))

**Downsides**: Requires you to go and install another program and you have to use a command line every time

**Recommended for**: Linux users, anyone who already has Python >=3.8 installed

### Package managers

Some package managers may include VI. However, there is no guarantee that they will be up to date all the time, as some may not be updated automatically.

Here's a list of package managers that VI is in and the operating systems they support:

- [Scoop](https://scoop.sh) (Windows) - `scoop bucket add games` then `scoop install vanilla-installer`
- `pip` as mentioned [above](#pypi)
- More coming soon! If you want to add one to a package manager you use, check the details out [here](CONTRIBUTING.md#packaging).

**Requirements**: The respective package manager installed

**Downsides**: You may have delayed updates of VI

**Recommended for**: Most Windows/macOS users that will use VI a lot but don't want to go through the extra pain of installing Python
-->
### Scripts (for advanced users)

These are usually not difficult to install as well. However, you may run into problems because of stability issues because it can always be used with the latest commit. Use at your own risk.

**[Windows](/install/windows.bat) · [MacOS](/install/macos.sh) · [Linux](/install/linux.sh)**

## Links

- **[Support](docs/support.md)**: Troubleshooting, compatibility notices & more
- **[Credits](docs/credits.md)**: Credits for the source code of VI
- **[Information for developers](docs/for-devs.md)**: Useful information for developers
- **[FO issue (#110)](https://github.com/Fabulously-Optimized/fabulously-optimized/issues/110)**: The issue in FO, which explains the situation

## Contributors

Thanks to all these wonderful people!
See [here](https://github.com/Fabulously-Optimized/vanilla-installer/graphs/contributors) for an overview of code contributors.
<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tbody>
    <tr>
      <td align="center" valign="top" width="14.28%"><a href="https://blobfox.coffee/@Kichura"><img src="https://avatars.githubusercontent.com/u/68134602?v=4?s=100" width="100px;" alt="Kichura"/><br /><sub><b>Kichura</b></sub></a><br /><a href="#infra-Kichura" title="Infrastructure (Hosting, Build-Tools, etc)">🚇</a></td>
      <td align="center" valign="top" width="14.28%"><a href="https://onlix.me"><img src="https://avatars.githubusercontent.com/u/67185896?v=4?s=100" width="100px;" alt="Felix"/><br /><sub><b>Felix</b></sub></a><br /><a href="#maintenance-nsde" title="Maintenance">🚧</a> <a href="https://github.com/Fabulously-Optimized/vanilla-installer/commits?author=nsde" title="Code">💻</a> <a href="#ideas-nsde" title="Ideas, Planning, & Feedback">🤔</a> <a href="https://github.com/Fabulously-Optimized/vanilla-installer/pulls?q=is%3Apr+reviewed-by%3Ansde" title="Reviewed Pull Requests">👀</a></td>
      <td align="center" valign="top" width="14.28%"><a href="https://osfanbuff63.tech"><img src="https://avatars.githubusercontent.com/u/91388253?v=4?s=100" width="100px;" alt="osfanbuff63"/><br /><sub><b>osfanbuff63</b></sub></a><br /><a href="#maintenance-osfanbuff63" title="Maintenance">🚧</a> <a href="https://github.com/Fabulously-Optimized/vanilla-installer/commits?author=osfanbuff63" title="Code">💻</a> <a href="https://github.com/Fabulously-Optimized/vanilla-installer/pulls?q=is%3Apr+reviewed-by%3Aosfanbuff63" title="Reviewed Pull Requests">👀</a> <a href="#ideas-osfanbuff63" title="Ideas, Planning, & Feedback">🤔</a></td>
      <td align="center" valign="top" width="14.28%"><a href="https://github.com/RaptaG"><img src="https://avatars.githubusercontent.com/u/77157639?v=4?s=100" width="100px;" alt="RaptaG"/><br /><sub><b>RaptaG</b></sub></a><br /><a href="https://github.com/Fabulously-Optimized/vanilla-installer/commits?author=RaptaG" title="Code">💻</a></td>
      <td align="center" valign="top" width="14.28%"><a href="https://github.com/RozeFound"><img src="https://avatars.githubusercontent.com/u/68745888?v=4?s=100" width="100px;" alt="Project D.D."/><br /><sub><b>Project D.D.</b></sub></a><br /><a href="https://github.com/Fabulously-Optimized/vanilla-installer/commits?author=RozeFound" title="Code">💻</a></td>
      <td align="center" valign="top" width="14.28%"><a href="https://github.com/tuxinal"><img src="https://avatars.githubusercontent.com/u/24763016?v=4?s=100" width="100px;" alt="Tuxinal"/><br /><sub><b>Tuxinal</b></sub></a><br /><a href="#design-tuxinal" title="Design">🎨</a> <a href="https://github.com/Fabulously-Optimized/vanilla-installer/commits?author=tuxinal" title="Code">💻</a></td>
      <td align="center" valign="top" width="14.28%"><a href="https://github.com/Madis0"><img src="https://avatars.githubusercontent.com/u/8611110?v=4?s=100" width="100px;" alt="Madis0"/><br /><sub><b>Madis0</b></sub></a><br /><a href="#infra-Madis0" title="Infrastructure (Hosting, Build-Tools, etc)">🚇</a> <a href="https://github.com/Fabulously-Optimized/vanilla-installer/pulls?q=is%3Apr+reviewed-by%3AMadis0" title="Reviewed Pull Requests">👀</a> <a href="#ideas-Madis0" title="Ideas, Planning, & Feedback">🤔</a></td>
    </tr>
  </tbody>
</table>

<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->

<!-- ALL-CONTRIBUTORS-LIST:END -->

![Contribution Overview](https://orbit.onlix.me/contribview/Fabulously-Optimized/vanilla-installer#15/11/2022)


CHANGELOG.md
# Changelog

The latest changelog can be found on the [wiki](https://fabulously-optimized.gitbook.io/vanilla-installer/changelogs/), or in the [GitHub releases](https://github.com/Fabulously-Optimized/vanilla-installer/releases).


vanilla-installer/gui.py
"""Runs the GUI for VanillaInstaller."""
# IMPORTS
import webbrowser
import pathlib
from PySide6.QtCore import QCoreApplication, QRect, Qt, QRunnable, QThreadPool, Slot
from PySide6.QtGui import QFont, QIcon
from PySide6.QtWidgets import (
    QApplication,
    QComboBox,
    QLabel,
    QMainWindow,
    QPushButton,
    QTextEdit,
    QWidget,
    QGraphicsColorizeEffect,
    QFileDialog,
    QDialog,
    QDialogButtonBox,
)
from PySide6.QtSvgWidgets import QSvgWidget
import minecraft_launcher_lib as mll

# LOCAL
import main
import theme

# ARGUMENTS
FONT = "Inter"


def run() -> None:
    """Runs the GUI."""
    app = QApplication([])
    window = QMainWindow()
    ui = Ui_MainWindow()
    ui.setupUi(window)
    get_versions_worker = Worker(ui.addVersions)
    ui.threadpool.start(get_versions_worker)
    window.show()
    app.exec()


class Ui_MainWindow(object):
    def setupUi(self, MainWindow: QMainWindow) -> None:
        if not MainWindow.objectName():
            MainWindow.setObjectName("MainWindow")
        MainWindow.resize(600, 400)
        MainWindow.setMinimumSize(600, 400)

        self.threadpool = QThreadPool(MainWindow)
        self.installing = False

        font = QFont()
        font.setFamily(FONT)
        MainWindow.setFont(font)
        self.centralwidget = QWidget(MainWindow)
        self.centralwidget.setObjectName("centralwidget")
        fontSize24 = QFont()
        fontSize24.setPointSize(24)
        self.title = QLabel(self.centralwidget)
        self.title.setObjectName("title")
        self.title.setGeometry(QRect(137, 42, 326, 40))
        self.title.setFont(fontSize24)
        self.title.setAlignment(Qt.AlignCenter)
        self.subtitle = QLabel(self.centralwidget)
        self.subtitle.setObjectName("subtitle")
        self.subtitle.setGeometry(QRect(0, 100, 600, 30))
        fontSize15 = QFont()
        fontSize15.setPointSize(15)
        self.subtitle.setFont(fontSize15)
        self.subtitle.setAlignment(Qt.AlignCenter)
        self.installButton = QPushButton(self.centralwidget)
        self.installButton.setObjectName("installButton")
        self.installButton.setGeometry(QRect(225, 164, 150, 50))
        self.installButton.setFont(fontSize15)
        self.installButton.clicked.connect(
            lambda: self.threadpool.start(Worker(self.startInstall))
        )
        fontSize12 = QFont()
        fontSize12.setPointSize(12)
        self.versionSelector = QComboBox(self.centralwidget)
        self.versionSelector.setObjectName("versionSelector")
        self.versionSelector.setGeometry(QRect(326, 240, 98, 20))
        self.versionSelector.setFont(fontSize12)
        self.versionLabel = QLabel(self.centralwidget)
        self.versionLabel.setObjectName("versionLabel")
        self.versionLabel.setGeometry(QRect(175, 240, 149, 20))
        self.versionLabel.setFont(fontSize12)
        self.versionLabel.setAlignment(Qt.AlignCenter)
        self.locationLabel = QLabel(self.centralwidget)
        self.locationLabel.setObjectName("locationLabel")
        self.locationLabel.setGeometry(QRect(130, 300, 100, 20))
        self.locationLabel.setFont(fontSize12)
        self.locationLabel.setAlignment(Qt.AlignCenter)

        self.selectedLocation = QTextEdit(self.centralwidget)
        self.selectedLocation.setObjectName("selectedLocation")
        self.selectedLocation.setGeometry(QRect(236, 295, 190, 30))
        self.selectedLocation.setFont(fontSize12)
        self.selectedLocation.setLineWrapMode(QTextEdit.LineWrapMode.NoWrap)
        self.selectedLocation.setVerticalScrollBarPolicy(
            Qt.ScrollBarPolicy.ScrollBarAlwaysOff
        )
        self.selectedLocation.setHorizontalScrollBarPolicy(
            Qt.ScrollBarPolicy.ScrollBarAlwaysOff
        )
        self.selectedLocation.setText(main.get_dir())

        self.locationSelector = QPushButton(self.centralwidget)
        self.locationSelector.setObjectName("locationSelector")
        self.locationSelector.setGeometry(QRect(440, 295, 30, 30))
        self.locationSelector.clicked.connect(
            lambda: self.selectDirectory(self.centralwidget)
        )
        self.locationSelectorIcon = QSvgWidget(
            Ui_MainWindow.getAsset("folder.svg"), self.locationSelector
        )
        self.locationSelectorIcon.setGeometry(5, 5, 20, 20)

        self.infoButton = QPushButton(self.centralwidget)
        self.infoButton.setObjectName("infoButton")
        self.infoButton.setGeometry(QRect(10, 366, 84, 24))
        self.infoButton.setFlat(True)
        self.infoButton.clicked.connect(
            lambda: webbrowser.open(
                "https://github.com/Fabulously-Optimized/vanilla-installer/"
            )
        )
        self.infoButtonIcon = QSvgWidget(
            Ui_MainWindow.getAsset("github.svg"), self.infoButton
        )
        self.infoButtonIcon.setGeometry(0, 0, 24, 24)

        self.issuesButton = QPushButton(self.centralwidget)
        self.issuesButton.setObjectName("issuesButton")
        self.issuesButton.setGeometry(QRect(10, 332, 108, 24))
        self.issuesButton.setFlat(True)
        self.issuesButton.clicked.connect(
            lambda: webbrowser.open(
                "https://github.com/Fabulously-Optimized/vanilla-installer/issues"
            )
        )
        self.issuesButtonIcon = QSvgWidget(
            Ui_MainWindow.getAsset("flag.svg"), self.issuesButton
        )
        self.issuesButtonIcon.setGeometry(0, 0, 24, 24)

        self.themeToggle = QPushButton(self.centralwidget)
        self.themeToggle.setObjectName("themeToggle")
        self.themeToggle.setGeometry(QRect(476, 366, 114, 24))
        self.themeToggle.setFlat(True)
        self.themeToggle.clicked.connect(self.toggleTheme)
        self.themeToggleIcon = QSvgWidget(self.themeToggle)
        self.themeToggleIcon.setGeometry(90, 0, 24, 24)

        self.settingsButton = QPushButton(self.centralwidget)
        self.settingsButton.setObjectName("settingsButton")
        self.settingsButton.setGeometry(QRect(506, 332, 84, 24))
        self.settingsButton.setFlat(True)
        self.settingsButton.clicked.connect(self.openSettings)
        self.settingsButtonIcon = QSvgWidget(
            Ui_MainWindow.getAsset("settings.svg"), self.settingsButton
        )

        self.settingsButtonIcon.setGeometry(60, 0, 24, 24)

        self.reloadTheme()

        MainWindow.setCentralWidget(self.centralwidget)

        self.retranslateUi(MainWindow)

    def retranslateUi(self, MainWindow) -> None:
        MainWindow.setWindowTitle(
            QCoreApplication.translate("MainWindow", "Vanilla Installer", None)
        )
        self.title.setText(
            QCoreApplication.translate("MainWindow", "Fabulously Optimized", None)
        )
        self.subtitle.setText(
            QCoreApplication.translate("MainWindow", "Vanilla Installer", None)
        )
        self.installButton.setText(
            QCoreApplication.translate("MainWindow", "Install", None)
        )
        self.versionLabel.setText(
            QCoreApplication.translate("MainWindow", "Minecraft version:", None)
        )
        self.locationLabel.setText(
            QCoreApplication.translate("MainWindow", "Location:", None)
        )
        self.infoButton.setText(
            QCoreApplication.translate("MainWindow", "GitHub", None)
        )
        self.issuesButton.setText(
            QCoreApplication.translate("MainWindow", "Report bugs", None)
        )
        self.themeToggle.setText(
            QCoreApplication.translate("MainWindow", "Toggle theme", None)
        )
        self.settingsButton.setText(
            QCoreApplication.translate("MainWindow", "Settings", None)
        )

    def reloadTheme(self) -> None:
        loaded_theme = theme.load()
        self.centralwidget.setStyleSheet(
            f'[objectName^="centralwidget"] {{ background-color: { loaded_theme.get("base")} }}'
        )
        self.title.setStyleSheet(f'color: { loaded_theme.get("text")}')
        self.subtitle.setStyleSheet(f'color: { loaded_theme.get("subtitle") }')

        self.installButton.setStyleSheet(
            f'QPushButton {{ border: none; background: {loaded_theme.get("blue")}; color: {loaded_theme.get("base")}; border-radius: 5px; }}'
            f'QPushButton:hover {{ background: {loaded_theme.get("lavender")};}}'
            f'QPushButton:pressed {{ background: {loaded_theme.get("installbuttonpressed")};}}'
        )
        self.locationSelector.setStyleSheet(
            f'QPushButton {{ border: none;background: {loaded_theme.get("button")}; border-radius: 5px; }}'
            f'QPushButton:hover {{ background: {loaded_theme.get("buttonhovered")}; }}'
            f'QPushButton:pressed {{ background: {loaded_theme.get("buttonpressed")}; }}'
        )
        self.infoButton.setStyleSheet(
            f"QPushButton{{ color: #00000000 }}"
            f'QPushButton:hover {{ color: {loaded_theme.get("label")}; text-align: left; padding-left: 30px}}'
        )
        self.issuesButton.setStyleSheet(
            f"QPushButton{{ color: #00000000 }}"
            f'QPushButton:hover {{ color: {loaded_theme.get("label")}; text-align: left; padding-left: 30px}}'
        )
        self.themeToggle.setStyleSheet(
            f"QPushButton{{ color: #00000000 }}"
            f'QPushButton:hover {{ color: {loaded_theme.get("label")}; text-align: right; padding-right: 30px}}'
        )
        self.settingsButton.setStyleSheet(
            f"QPushButton{{ color: #00000000 }}"
            f'QPushButton:hover {{ color: {loaded_theme.get("label")}; text-align: right; padding-right: 30px}}'
        )

        self.versionLabel.setStyleSheet(f'color: {loaded_theme.get("label")}')
        self.locationLabel.setStyleSheet(f'color: {loaded_theme.get("label")}')
        self.selectedLocation.setStyleSheet(
            f'color: {loaded_theme.get("text")}; background-color: {loaded_theme.get("crust")}'
        )
        self.themeToggleIcon.load(
            Ui_MainWindow.getAsset("moon.svg" if theme.is_dark() else "sun.svg")
        )
        # each effect can only apply to one widget
        effect1 = QGraphicsColorizeEffect(self.centralwidget)
        effect1.setColor(loaded_theme.get("icon"))
        self.locationSelectorIcon.setGraphicsEffect(effect1)
        effect2 = QGraphicsColorizeEffect(self.centralwidget)
        effect2.setColor(loaded_theme.get("icon"))
        self.infoButtonIcon.setGraphicsEffect(effect2)
        effect3 = QGraphicsColorizeEffect(self.centralwidget)
        effect3.setColor(loaded_theme.get("icon"))
        self.issuesButtonIcon.setGraphicsEffect(effect3)
        effect4 = QGraphicsColorizeEffect(self.centralwidget)
        effect4.setColor(loaded_theme.get("icon"))
        self.themeToggleIcon.setGraphicsEffect(effect4)
        effect5 = QGraphicsColorizeEffect(self.centralwidget)
        effect5.setColor(loaded_theme.get("icon"))
        self.settingsButtonIcon.setGraphicsEffect(effect5)

    def addVersions(self) -> None:
        for version in main.get_pack_mc_versions():
            self.versionSelector.addItem(version)
        self.versionSelector.setCurrentIndex(0)

    def getAsset(asset) -> str:
        return str((pathlib.Path(__file__).parent / "assets" / asset).resolve())

    def toggleTheme(self) -> None:
        theme.toggle()
        self.reloadTheme()

    def selectDirectory(self, parent) -> None:
        dialog = QFileDialog(
            parent,
            QCoreApplication.translate("MainWindow", "Select .minecraft folder", None),
        )
        dialog.setFileMode(QFileDialog.FileMode.Directory)
        current_path = pathlib.Path(self.selectedLocation.toPlainText()).absolute()
        if not current_path.exists():
            current_path = pathlib.Path(mll.utils.get_minecraft_directory()).absolute()
        dialog.setDirectory(str(current_path))
        if dialog.exec():
            self.selectedLocation.setText(dialog.selectedFiles()[0])

    def openSettings(self) -> None:
        dialog = SettingsDialog(self.centralwidget)
        dialog.exec()

    def startInstall(self) -> None:
        # make sure the installation process is only running once
        if self.installing is True:
            return
        version = self.versionSelector.itemText(self.versionSelector.currentIndex())
        location = self.selectedLocation.toPlainText()
        self.installing = True
        main.run(self.subtitle, location, version)
        self.installing = False


class SettingsDialog(QDialog):
    def __init__(self, parent) -> None:
        super().__init__(parent)
        self.setupUi()

    def setupUi(self) -> None:
        if not self.objectName():
            self.setObjectName("Dialog")
        self.setMinimumSize(400, 250)
        font = QFont()
        font.setFamily(FONT)
        self.setFont(font)
        self.buttonBox = QDialogButtonBox(self)
        self.buttonBox.setObjectName("buttonBox")
        self.buttonBox.setGeometry(QRect(180, 190, 200, 40))
        self.buttonBox.setOrientation(Qt.Horizontal)
        self.buttonBox.setStandardButtons(
            QDialogButtonBox.Cancel | QDialogButtonBox.Save
        )
        self.buttonBox.setCenterButtons(False)

        user = key = None
        auth = main.get_gh_auth()
        if auth is not None:
            user, key = auth
        self.userLabel = QLabel(self)
        self.userLabel.setObjectName("userLabel")
        self.userLabel.setGeometry(QRect(20, 20, 105, 20))

        self.userInput = QTextEdit(self)
        self.userInput.setObjectName("userInput")
        self.userInput.setGeometry(QRect(140, 15, 240, 30))
        self.userInput.setLineWrapMode(QTextEdit.LineWrapMode.NoWrap)
        self.userInput.setVerticalScrollBarPolicy(Qt.ScrollBarPolicy.ScrollBarAlwaysOff)
        self.userInput.setHorizontalScrollBarPolicy(
            Qt.ScrollBarPolicy.ScrollBarAlwaysOff
        )
        self.userInput.setText(user)
        self.userInput.setTabChangesFocus(True)
        self.keyLabel = QLabel(self)
        self.keyLabel.setObjectName("keyLabel")
        self.keyLabel.setGeometry(QRect(20, 60, 105, 20))
        self.keyInput = QTextEdit(self)
        self.keyInput.setObjectName("keyInput")
        self.keyInput.setGeometry(QRect(140, 55, 240, 30))
        self.keyInput.setLineWrapMode(QTextEdit.LineWrapMode.NoWrap)
        self.keyInput.setVerticalScrollBarPolicy(Qt.ScrollBarPolicy.ScrollBarAlwaysOff)
        self.keyInput.setHorizontalScrollBarPolicy(
            Qt.ScrollBarPolicy.ScrollBarAlwaysOff
        )
        self.keyInput.setText(key)
        self.keyInput.setTabChangesFocus(True)
        for button in self.buttonBox.buttons():
            button.setIcon(QIcon())  # remove the button icons
        fontSize8 = QFont()
        fontSize8.setPointSize(8)
        self.loginInfoLabel = QLabel(self)
        self.loginInfoLabel.setFont(fontSize8)
        self.loginInfoLabel.setWordWrap(True)
        self.loginInfoLabel.setGeometry(QRect(20, 90, 360, 70))

        self.errorLabel = QLabel(self)
        self.errorLabel.setFont(fontSize8)
        self.errorLabel.setWordWrap(True)
        self.errorLabel.setGeometry(QRect(20, 200, 200, 20))
        self.reloadTheme()
        self.retranslateUi(self)
        self.buttonBox.accepted.connect(self.accept)
        self.buttonBox.rejected.connect(self.reject)

    def accept(self) -> None:
        status = main.set_gh_auth(
            "".join(self.userInput.toPlainText().split()),
            "".join(self.keyInput.toPlainText().split()),
        )  # strips whitespace
        if status == False:
            self.errorLabel.setText(
                QCoreApplication.translate("Dialog", "Invalid login details!", None)
            )
            return
        super().accept()

    def retranslateUi(self, Dialog) -> None:
        Dialog.setWindowTitle(
            QCoreApplication.translate("Dialog", "Vanilla Installer Settings", None)
        )
        self.keyLabel.setText(
            QCoreApplication.translate("Dialog", "Github API Token:", None)
        )
        self.userLabel.setText(
            QCoreApplication.translate("Dialog", "Github Username:", None)
        )
        self.loginInfoLabel.setText(
            QCoreApplication.translate(
                "Dialog",
                "Set your GitHub token to a personal token in order to extend your API ratelimit\n If you don't know what that is and aren't going to install Fabulously Optimized more than 60 times in an hour, You probably don't need this",
                None,
            )
        )

    def reloadTheme(self) -> None:
        loaded_theme = theme.load()
        self.setStyleSheet(
            f'[objectName^="Dialog"] {{ background-color: {loaded_theme.get("base")}}}'
        )
        self.buttonBox.setStyleSheet(
            f'QPushButton {{ border: none; background-color: { loaded_theme.get("button") } ; color: {loaded_theme.get("text")}; padding: 8px; border-radius: 5px}}'
            f'QPushButton:hover {{ background-color: { loaded_theme.get("buttonhovered") }}}'
            f'QPushButton:hover {{ background-color: { loaded_theme.get("buttonpressed") }}}'
        )
        self.keyLabel.setStyleSheet(f'color: {loaded_theme.get("label")}')
        self.keyInput.setStyleSheet(
            f'background-color: {loaded_theme.get("crust")}; color: {loaded_theme.get("text")};'
        )
        self.userLabel.setStyleSheet(f'color: {loaded_theme.get("label")}')
        self.userInput.setStyleSheet(
            f'background-color: {loaded_theme.get("crust")}; color: {loaded_theme.get("text")};'
        )
        self.loginInfoLabel.setStyleSheet(f'color: {loaded_theme.get("subtitle")}')
        self.errorLabel.setStyleSheet(f'color: {loaded_theme.get("red")}')


class Worker(QRunnable):
    def __init__(self, fn) -> None:
        super(Worker, self).__init__()
        self.fn = fn

    @Slot()
    def run(self):
        self.fn()


if __name__ == "__main__":
    run()
